File: scala/com/huawei/graphblas/examples/Pagerank.scala

/*
 *   Copyright 2021 Huawei Technologies Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package com.huawei.graphblas.examples

import sys.process._

import org.apache.spark.SparkEnv
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

import com.huawei.GraphBLAS

object Pagerank {

	def spark_pr( matrix: org.apache.spark.rdd.RDD[ (Int, Iterable[Int] ) ], iters: Int, cp_freq: Int = 30 ): org.apache.spark.rdd.RDD[ (Int, Double) ] = {
		var ranks: org.apache.spark.rdd.RDD[ (Int,Double) ] = matrix.map( x => (x._1, 1.0 ) );
		for( i <- 1 to iters ) {
			if( (i % (cp_freq+1)) == cp_freq ) {
				ranks.persist();
			}
			ranks = matrix.join(ranks).map( x => x._2 ).flatMap( x => x._1.map( y => (y, 0.15 + 0.85 * x._2) ) ).reduceByKey( _ + _ );
		}
		ranks
	}

	def main( args: Array[String] ): Unit = {
		if( args.length != 2 ) {
			println( "Usage: ./program_name <P> <matrix file>." );
			return;
		}
		val sc = new SparkContext( new SparkConf().setAppName( "Spark GraphBLAS Pagerank" ) );
		val P = args(0).toInt;
		println( s"Using P = $P." );
		val t0 = System.nanoTime();
		val hostnames = sc.parallelize( 0 until P ).map{ pid => {(SparkEnv.get.executorId,"hostname"!!)} }.collect().toArray
		println("Manual hostnames gathering at start of example:")
		println(hostnames.deep)
		println("Now creating GraphBLAS launcher:")
		val grb = com.huawei.GraphBLAS.initialize( sc, P )
		println("grb instance contents:")
		println(grb)
		val t1 = System.nanoTime();
		val output = com.huawei.GraphBLAS.pagerank( sc, grb, args(1) );
		val t2 = System.nanoTime();
		println("output contents:")
		println(output)
		val t3 = System.nanoTime();
		val maxpair = com.huawei.GraphBLAS.max( sc, grb, output );
		val t4 = System.nanoTime();
		println("maximum PageRank entry:");
		println(maxpair);
		com.huawei.GraphBLAS.destroy( sc, grb, output );
		com.huawei.GraphBLAS.exit( sc, grb )
		println("GraphBLAS cleaned up.")
		val time_taken = (System.nanoTime() - t0) / 1000000000.0;
		val t21 = (t2 - t1) / 1000000000.0;
		val t43 = (t4 - t3) / 1000000000.0;
		println( s"Accelerated PageRank call: $t21 seconds." );
		println( s"Max extraction call: $t43 seconds." );
		println( s"End-to-end time taken: $time_taken seconds." );
	}
}

File: cpp/pagerank.cpp

/*
 *   Copyright 2021 Huawei Technologies Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string>
#include <stdlib.h>
#include <chrono>
#include <limits>
#include <vector>

#include "graphblas/algorithms/simple_pagerank.hpp"
#include "graphblas/utils/parser/MatrixFileReader.hpp"

#include "sparkgrb.hpp"

#include "pagerank.hpp"
#include "logger.h"


using duration_t = std::chrono::time_point<std::chrono::high_resolution_clock>::duration;


static std::vector< double > ms_times;

double* get_ms_times_pointer() {
	return ms_times.data();
}

size_t get_ms_times_size() {
	return ms_times.size();
}


static std::vector< unsigned > iterations;

unsigned* get_iterations_pointer() {
	return iterations.data();
}

size_t get_iterations_size() {
	return iterations.size();
}


void do_pagerank( const pagerank_input &data_in, pagerank_output &out ) {

	try {
		const std::size_t n = grb::nrows( *data_in.data );
		grb::Vector< double > pr( n ), workspace1( n ), workspace2( n ), workspace3( n );
		LOG( INFO, "Output vector allocated, now passing to PageRank function" );
		out.error_code = grb::algorithms::simple_pagerank< grb::descriptors::no_operation >(
			pr, *data_in.data,
			workspace1, workspace2, workspace3,
			data_in.alpha, data_in.tolerance, data_in.max_iterations,
			&( out.iterations ), &( out.residual )
		);

		if( out.error_code != grb::SUCCESS ) {
			LOG( ERROR, "error running PageRank warmup" );
			return;
		}

		// duration_t pr_time = duration_t::zero();
		ms_times.clear();
		iterations.clear();
		LOG( INFO, "available threads: %d", omp_get_max_threads() );
		LOG( INFO, "PageRank running for %u times", data_in.outer_iters );
		for( unsigned i = 0; i < data_in.outer_iters; i++ ) {
			grb::clear( pr );
			auto start = std::chrono::high_resolution_clock::now();
			out.error_code = grb::algorithms::simple_pagerank< grb::descriptors::no_operation >(
				pr, *data_in.data,
				workspace1, workspace2, workspace3,
				data_in.alpha, data_in.tolerance, data_in.max_iterations,
				&( out.iterations ), &( out.residual )
			);
			auto end = std::chrono::high_resolution_clock::now();
			ms_times.push_back( std::chrono::duration_cast< std::chrono::duration< double, std::milli > >( end - start ).count() );
			iterations.push_back( out.iterations );

			LOG( INFO, "PageRank ran for %u iterations, %lf ms",
				iterations.back(), ms_times.back() );
		}

		if( out.error_code != grb::SUCCESS ) {
			std::string error_code = grb::toString( out.error_code );
			LOG( ERROR, "Call to PageRank failed with error code %s; ", error_code.c_str() );
			return;
		}
		LOG( INFO, "Call to PageRank successful; " );
		out.pinnedVector = new grb::PinnedVector< double >( pr, grb::PARALLEL );

		LOG( INFO, "iters = %zd, residual = %.10e", out.iterations, out.residual );
		LOG( INFO, "returning pinned vector @ %p. It contains %zd elements.", out.pinnedVector, out.pinnedVector->nonzeroes() );
	} catch( std::runtime_error &e ) {
		LOG( INFO, "Got exception: %s", e.what() );
		out.error_code = grb::PANIC;
	}
	LOG( INFO, "Exiting grb_program." );
}

/**
 * Starts the GraphBLAS pagerank algorithm in #grb::algorithms by reading an
 * input file in parallel.
 *
 * @param[in]  data_in  The input structure. Contains the file name of the input
 *                      matrix.
 * @param[out] data_out The output pagerank vector.
 */
void grb_pagerank_from_file( const pagerank_file_input &data_in, pagerank_output &out ) {
	out.error_code = grb::PANIC;
	out.iterations = std::numeric_limits< size_t >::max();
	out.residual = std::numeric_limits< double >::infinity();
	out.pinnedVector = nullptr;

	LOG( INFO, "This is process %d. I am BSP process %lu out of %lu. OpenMP uses %d "
		"threads. My OMP_NUM_THREADS %s.", getpid(), grb::spmd<>::pid(),
		grb::spmd<>::nprocs(), omp_get_max_threads(), getenv( "OMP_NUM_THREADS" ) );

	LOG( INFO, "I will perform a PageRank using GraphBLAS I/O. Filename: %s", data_in.infile );

	try {
		grb::utils::MatrixFileReader< void > parser( data_in.infile, true );

		const size_t m = parser.m();
		const size_t n = parser.n();
		if( m != n ) {
			out.error_code = grb::ILLEGAL;
			return;
		}
		grb::Matrix< void > L( n, n );
		out.error_code = grb::buildMatrixUnique( L, parser.begin( grb::PARALLEL ), parser.end( grb::PARALLEL ), grb::PARALLEL );
		if( out.error_code != grb::SUCCESS ) {
			LOG( ERROR, "cannot read input using GraphBLAS I/O. Filename: %s", data_in.infile );
			return;
		}
		try {
			if( grb::nnz( L ) != parser.nz() ) {
				LOG( ERROR,
					"Error: number of nonzeroes in grb::Matrix (%zd) does not match that "
					"reported by the parser (%zd)", grb::nnz( L ), parser.nz()
				);
				out.error_code = grb::PANIC;
				return;
			}
		} catch( ... ) {
			LOG( WARNING,
				"exception caught during call involing parser.nz(). Assuming that "
				"this is cased by input matrix symmetry."
			);
		}
		pagerank_input final( data_in );
		final.data = &L;

		do_pagerank( final, out );

	} catch( std::runtime_error &e ){
		LOG( ERROR, "Got exception: %s", e.what() );
		out.error_code = grb::PANIC;
	}

	LOG( INFO, "Exiting grb_pagerank_from_file." );
	LOG_EXIT();
}
